### ISSUE TITLE
Video play dispose and loading issue

### ISSUE THREAD

**Issue by Minsamin** [darwin-morocho/flutter-meedu-player#33]
If I use autoplay false, the loading continues.
If i tap on 10Sec buttons or loading widget, then playButton appears.
Using
meedu_player: ^0.4.1+1

One more issue,
If i frequently toggle full screen On Off, then the dispose error showing -> 

A VideoPlayerController was used after being disposed.

Using same code from [https://player.meedu.app/#/?id=how-to-use](url)

**Comment by darwin-morocho** [comment]
> https://player.meedu.app/#/?id=how-to-use

Hi @Minsamin, Are  you trying to have multiples videos inside a ListView.Builder>

**Comment by Minsamin** [comment]
I am creating a feeds pagez so it contains images, or text, or videos.
So, listview items, can contain, single video.

 ![Screenshot_20210108-205741.jpg](https://user-images.githubusercontent.com/23526722/104032441-2a640480-51f4-11eb-959f-8db5dd1e849c.jpg)

This is the screen, and the loading forever, until I press 10Sec buttons.

**Comment by Minsamin** [comment]
Screen Recording of the issue.

**Comment by Minsamin** [comment]
https://user-images.githubusercontent.com/23526722/104033388-75cae280-51f5-11eb-977b-39196f46b386.mp4

Screen Recording of the issue.

**Comment by darwin-morocho** [comment]
>  20210108_210646.mp4 
> Screen Recording of the issue.

The problem is because the ListView.builder  destroy the widgets when they are not visible.  To avoid this maybe you need use a global key for each item in your list. I will try to find a good solution for your case and add a example in the example folder.

**Comment by darwin-morocho** [comment]
@Minsamin the loading animation was fixed on the version **0.4.2**.


For the dispose error please can you share the code or let me know if using globalKey the error is solved

**Comment by Minsamin** [comment]
Sure sir, I will update you, after I implement your suggestion. Thanks...

**Comment by Minsamin** [comment]
Loading Issue Fixed in version 0.4.2. Thanks
But Full-screen toggle shows error some time.


https://user-images.githubusercontent.com/23526722/104149198-4ab1e000-53fb-11eb-8fe0-a7c861c9bb7f.mp4



I am using sample Stateful Widget code.

![image](https://user-images.githubusercontent.com/23526722/104149312-b8f6a280-53fb-11eb-8121-825a9835764f.png)

**Comment by Minsamin** [comment]
Let me know if I'm using Global key correct way! Cause this not solves the issue. Same error.

This is my PageList Library for pagination in Feed Screen. This is a Stateless Widget
![image](https://user-images.githubusercontent.com/23526722/104157706-b30bbc00-5411-11eb-9c57-2ccfa9c72e36.png)


This is List Item Stateful Widget
![image](https://user-images.githubusercontent.com/23526722/104157854-05e57380-5412-11eb-8c31-1d4d45b9c170.png)

And Here I can have Images or Videos or text in List View Item. So I have if conditions and show image or Stateful Video Widget (Code screenshot given in above comment).

**Comment by darwin-morocho** [comment]
> @darwin-morocho , Let me know if I'm using Global key correct way! Cause this not solves the issue. Same error.
> 
> This is my PageList Library for pagination in Feed Screen. This is a Stateless Widget
> ![image](https://user-images.githubusercontent.com/23526722/104157706-b30bbc00-5411-11eb-9c57-2ccfa9c72e36.png)
> 
> This is List Item Stateful Widget
> ![image](https://user-images.githubusercontent.com/23526722/104157854-05e57380-5412-11eb-8c31-1d4d45b9c170.png)
> 
> And Here I can have Images or Videos or text in List View Item. So I have if conditions and show image or Stateful Video Widget (Code screenshot given in above comment).

Keys help you to preserve the widget state when you have a widget inside a list or when the ui changes for screen rotation.
For that reason you need create a key for each item in your list, and store these keys into a property list

**Comment by Minsamin** [comment]
Very Sorry, cant understand, any snippet or example code is highly appreciable.

**Comment by Minsamin** [comment]
![image](https://user-images.githubusercontent.com/23526722/104440320-32d38b00-55b8-11eb-8b4c-864491cb1796.png)

I have used this two... But both has the same error.

I think it's not the keys, because if list view has only one video showing, then there is no problem.
Only if two video is fully visible in the screen, and I click on Full screen button, then I got this error.

 
![image](https://user-images.githubusercontent.com/23526722/104440920-f8b6b900-55b8-11eb-847a-ca6a120ab7fa.png)

**Comment by Minsamin** [comment]
"Only if two video is fully visible in the screen, and I click on the Full screen button, then I got this error."


https://user-images.githubusercontent.com/23526722/104446306-4387ff00-55c0-11eb-90c3-592e9e47d057.mp4

**Comment by darwin-morocho** [comment]
@Minsamin I been tried to reproduce your error and yes the problem is because the ListView.builder destroy the element when the device rotate. I could fix it using AutomaticKeepAliveClientMixin but each item in the list never dies when you do scroll so you need to be careful with the performance

```dart 
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:meedu_player/meedu_player.dart';

class ListViewExample extends StatefulWidget {
  ListViewExample({Key key}) : super(key: key);

  @override
  _ListViewExampleState createState() => _ListViewExampleState();
}

class _ListViewExampleState extends State<ListViewExample> with AutomaticKeepAliveClientMixin {
  // final List<MeeduPlayerController> _controllers = [];

  // @override
  // void dispose() {
  //   _controllers.forEach((element) {
  //     element.dispose();
  //   });
  //   super.dispose();
  // }

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    super.build(context);
    return Scaffold(
      appBar: AppBar(),
      body: ListView.builder(
        cacheExtent: 600,
        itemBuilder: (_, index) => VideoItem(),
        itemCount: 10,
      ),
    );
  }

  @override
  bool get wantKeepAlive => true;
}

class VideoItem extends StatefulWidget {
  VideoItem({Key key}) : super(key: key);

  @override
  _VideoItemState createState() => _VideoItemState();
}

class _VideoItemState extends State<VideoItem> with AutomaticKeepAliveClientMixin {
  MeeduPlayerController _controller = MeeduPlayerController(
    screenManager: ScreenManager(orientations: [
      DeviceOrientation.portraitUp,
    ]),
  );

  @override
  void initState() {
    super.initState();
    _controller.setDataSource(
      DataSource(
        source: 'https://www.radiantmediaplayer.com/media/big-buck-bunny-360p.mp4',
        type: DataSourceType.network,
      ),
      autoplay: false,
    );
  }

  @override
  void dispose() {
    _controller.dispose();
    print("❌ dispose video player");
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    super.build(context);
    return AspectRatio(
      aspectRatio: 16 / 9,
      child: MeeduVideoPlayer(
        controller: _controller,
      ),
    );
  }

  @override
  bool get wantKeepAlive => true;
}

```

**Comment by Minsamin** [comment]
It's a pagination, so I can't keep all items in memory, this list can be more than 5k, maybe more in long term of the project. How can i use this video library then?
Can't you find any other solution?
Because I want to use this package, it's awesome. But with this issue i can deliver in production.

Please try to find some other ways. Is this issue has every video package.? I think not.

**Comment by darwin-morocho** [comment]
> It's a pagination, so I can't keep all items in memory, this list can be more than 5k, maybe more in long term of the project. How can i use this video library then?
> Can't you find any other solution?
> Because I want to use this package, it's awesome. But with this issue i can deliver in production.
> 
> Please try to find some other ways. Is this issue has every video package.? I think not.

@Minsamin this is not an issue is a normal behavior of this and other plugin, here you need to find a solution like destroy the widgets when they are no visible or something similar.

**Comment by Minsamin** [comment]
Ookk, waiting for your solution.

**Comment by darwin-morocho** [comment]
> Ookk, waiting for your solution.

you can find more info here https://github.com/flutter/flutter/issues/58917

And you could try this package https://pub.dev/packages/visibility_detector

**Comment by darwin-morocho** [comment]
@Minsamin other alternative is you can show a preview in your list as an image preview and when you click  the image show the player in fullscreen mode

**Comment by Minsamin** [comment]
> @Minsamin other alternative is you can show a preview in your list as an image preview and when you click  the image show the player in fullscreen mode

Yeah, I'm already using this. Not showing full screen button in list. Only full screen will work in feed details screen.
Thanks BTW. 😊😊
